### PR TITLE
[Feature] Supporting Tversky loss

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation

Adding new loss function named `Tversky Loss`. https://arxiv.org/abs/1706.05721

Old PR: https://github.com/open-mmlab/mmsegmentation/pull/634

## Modification

Adding `loss_name` which may be useful for multi losses training.

Note: Not too much difference with original dice_loss.py, just adding weight on false positives and false negatives.

|    Datasets    | loss_ce (from [repo config](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/unet)) | loss_ce + loss_tversky (1:1) | loss_ce + loss_tversky (1:3) | loss_tversky |
| ---------- | --- | --- | --- | --- |
| DRIVE |  <u> 78.67 </u> | 79.58 | 80.08 | **79.78** |
| STARE       |  81.02 | 81.97 | **82.74** | 82.36 |
| CHASE_DB1       |  80.24 | **80.36** | 80.1 | 77.49 |
| HRF       |  79.45 | **81.16** | 80.98 | 78.69 |

### CODE REVIEW

File: mmseg/models/losses/tversky_loss.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+
+@weighted_loss
+def binary_tversky_loss(pred,
+                        target,
+                        valid_mask,
+                        smooth=1,
+                        alpha=0.3,
+                        beta=0.7,
+                        **kwards):
```
Comment (xvjiarui) [inline mmseg/models/losses/tversky_loss.py:49]: Is `kwards` useful?

File: mmseg/models/losses/tversky_loss.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        self.alpha = alpha
+        self.beta = beta
+        self._loss_name = loss_name
+
+    def forward(self,
+                pred,
+                target,
+                avg_factor=None,
+                reduction_override=None,
+                **kwards):
```
Comment (xvjiarui) [inline mmseg/models/losses/tversky_loss.py:121]: Is kwards useful?

File: mmseg/models/losses/tversky_loss.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                 smooth=1,
+                 exponent=2,
+                 reduction='mean',
+                 class_weight=None,
+                 loss_weight=1.0,
+                 ignore_index=255,
+                 alpha=0.3,
+                 beta=0.7,
+                 loss_name='loss_tversky',
+                 **kwards):
```
Comment (xvjiarui) [inline mmseg/models/losses/tversky_loss.py:104]: Is kwards useful?

File: mmseg/models/losses/tversky_loss.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+@LOSSES.register_module()
+class TverskyLoss(nn.Module):
+    """TverskyLoss. This loss is proposed in `Tversky loss function for image
+    segmentation using 3D fully convolutional deep networks.
+
+    <https://arxiv.org/abs/1706.05721>`_.
+
+    Args:
+        loss_type (str, optional): Binary or multi-class loss.
+            Default: 'multi_class'. Options are "binary" and "multi_class".
```
Comment (xvjiarui) [inline mmseg/models/losses/tversky_loss.py:72]: Docstring not correct.

File: mmseg/models/losses/tversky_loss.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_tversky'.
+    """
+
+    def __init__(self,
+                 smooth=1,
+                 reduction='mean',
+                 class_weight=None,
+                 loss_weight=1.0,
+                 ignore_index=255,
```
Comment (xvjiarui) [inline mmseg/models/losses/tversky_loss.py:92]: Please remove this.

Comment (xvjiarui) [reply mmseg/models/losses/tversky_loss.py:92]: Also change dice_loss accordingly. 
`ignore_index` should be passed in forward not `__init__`.

File: mmseg/models/losses/tversky_loss.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                pred[:, i],
+                one_hot_target[..., i],
+                valid_mask=valid_mask,
+                smooth=smooth,
+                alpha=alpha,
+                beta=beta)
+            if class_weight is not None:
+                tversky_loss *= class_weight[i]
+            total_loss += tversky_loss
+    return total_loss / num_classes
```
Comment (xvjiarui) [inline mmseg/models/losses/tversky_loss.py:41]: We may refactor the for loop into vectorized implementation next month

**Comment by RockeyCoss** [comment]
Please merge the master branch into your branch, thank you.

**Comment by iumyx2612** [comment]
What happened to this PR?

**Comment by MengzhangLI** [comment]
> What happened to this PR?

It has been merged: https://github.com/open-mmlab/mmsegmentation/pull/1986. Sorry for your waiting for a long time.
